rows_test.go

```diff
@@ -75,3 +77,285 @@ func TestRowsClose(t *testing.T) {
+	if err != nil {
+		t.Fatalf("opening db failed: %s", err)
+	}
+
+	ctx := context.Background()
+	ctxKey := "ctxkey"
+	ctxVal := "1"
+
+	ctx = context.WithValue(ctx, ctxKey, ctxVal) // nolint: staticcheck // not using a custom type for the ctx key is not an issue here
+
+	qrs, err := db.QueryContext(ctx, "", "")
+	if err != nil {
+		t.Fatalf("db.Query failed: %s", err)
+	}
+
+	names, err := qrs.Columns()
+	if err != nil {
+		t.Errorf("error calling Columns, %v", err)
+	}
+
+	cts, err := qrs.ColumnTypes()
+	if err != nil {
+		t.Errorf("error calling ColumnTypes, %v", err)
+	}
+
+	if len(names) != 2 || len(names) != len(cts) {
+		t.Errorf("wrong column name or types count")
+	}
+
+	// There's no real way these can be called, but we'll text incase the
+	// test implementation changes
+	if rs.hasNextResultSetCalled {
+		t.Errorf("HasNextResultSetCalled, on non-supporting type, %v", err)
+	}
+	if rs.nextResultSetCalled {
+		t.Errorf("NextResultSetCalled, on non-supporting type, %v", err)
+	}
+
+	if !rs.columnTypeDatabaseNameCalled {
+		t.Errorf("ColumnTypeDatabaseName not called, %v", err)
+	}
+
+	if !rs.columnTypeLengthCalled {
+		t.Errorf("ColumnTypeTypeLenght not called, %v", err)
+	}
+
+	if !rs.columnTypeNullable {
+		t.Errorf("ColumnTypeTypeLenght not called, %v", err)
+	}
+
+	if !rs.columnTypePrecisionScaleCalled {
+		t.Errorf("ColumnTypePrecisionScale not called, %v", err)
+	}
+
+	var id, name string
+	for qrs.Next() {
+		err := qrs.Scan(&id, &name)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	err = qrs.Close()
+	if err != nil {
+		t.Errorf("rows Close failed: %s", err)
+	}
+
+	if !rows.nextCalled {
+		t.Error("driver rows.Next was not called")
+	}
+
+	if !interceptor.rowsNextCalled {
+		t.Error("interceptor rows.Next was not called")
+	}
+
+	if !interceptor.rowsNextCalled {
+		t.Error("interceptor rows.Next was not called")
+	}
+
+	if interceptor.rowsNextLastCtx == nil {
+		t.Fatal("rows close ctx is nil")
+	}
+
+	v := interceptor.rowsNextLastCtx.Value(ctxKey)
+	if v == nil {
+		t.Fatalf("ctx is different, missing value for key: %s", ctxKey)
+	}
+
+	vStr, ok := v.(string)
+	if !ok {
+		t.Fatalf("ctx is different, value for key: %s, has type %t, expected string", ctxKey, v)
+	}
+
+	if ctxVal != vStr {
+		t.Errorf("ctx is different, value for key: %s, is %q, expected %q", ctxKey, vStr, ctxVal)
+	}
+}
+
+func TestWrapRows(t *testing.T) {
+	ctx := context.Background()
+	tt := []struct {
+		name string
+		rows driver.Rows
+	}{
+		{
+			name: "vanilla",
+			rows: &fakeRows{},
+		},
+		{
+			name: "pgx",
+			rows: &fakeRowsLikePgx{},
+		},
+		{
+			name: "mysql",
+			rows: &fakeRowsLikeMysql{},
+		},
+	}
+
+	for _, st := range tt {
+		st := st
+		t.Run(st.name, func(t *testing.T) {
+			rows := st.rows
+			wr := wrapRows(ctx, nil, rows)
+
+			_, rok := rows.(driver.RowsNextResultSet)
+			_, wok := wr.(driver.RowsNextResultSet)
+			if rok != wok {
+				t.Fatalf("inconsinstent support for driver.RowsNextResultSet")
```

Comment:
optional:

```suggestion
				t.Fatalf("inconsistent support for driver.RowsNextResultSet")
```
That typo occurs multiple times. :-)